File: single-number/single_number_test.go
package singleNumber

import "testing"

func TestSingleNumber(t *testing.T) {
	var testCases = []struct {
		nums []int
		want int
	}{
		{[]int{2, 2, 1}, 1},
		{[]int{4, 1, 2, 1, 2}, 4},
		{[]int{1}, 1},
	}

	for _, tc := range testCases {
		if r := singleNumber2(tc.nums); r != tc.want {
			t.Errorf("singleNumber(%v) = %d, expected %d", tc.nums, r, tc.want)
		}
	}
}

File: top-interview/array/single-number/python/single_number_test.py
import unittest

from single_number import Solution

class TestSingleNumber(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()
        self.testCases = [
            {
                'nums': [2,2,1],
                'want': 1,
            },
            {
                'nums': [4,1,2,1,2],
                'want': 4,
            },
            {
                'nums': [1],
                'want': 1,
            },
        ]

    def test(self):
        for tc in self.testCases:
            with self.subTest(testCase=tc):
                self.assertEqual(
                    self.solution.singleNumber(tc['nums']),
                    tc['want'],
                )


if __name__ == '__main__':
    unittest.main()
